test/asynchronous/__init__.py

```diff
@@ -1053,3 +1113,95 @@ def print_running_clients():
                 processed.add(obj._topology_id)
         except ReferenceError:
             pass
+
+
+async def _all_users(db):
+    return {u["user"] for u in (await db.command("usersInfo")).get("users", [])}
+
+
+async def _create_user(authdb, user, pwd=None, roles=None, **kwargs):
+    cmd = SON([("createUser", user)])
+    # X509 doesn't use a password
+    if pwd:
+        cmd["pwd"] = pwd
+    cmd["roles"] = roles or ["root"]
+    cmd.update(**kwargs)
+    return await authdb.command(cmd)
+
+
+async def connected(client):
+    """Convenience to wait for a newly-constructed client to connect."""
+    with warnings.catch_warnings():
+        # Ignore warning that ping is always routed to primary even
+        # if client's read preference isn't PRIMARY.
+        warnings.simplefilter("ignore", UserWarning)
+        await client.admin.command("ping")  # Force connection.
+
+    return client
+
+
+async def drop_collections(db: AsyncDatabase):
+    # Drop all non-system collections in this database.
+    for coll in await db.list_collection_names(filter={"name": {"$regex": r"^(?!system\.)"}}):
+        await db.drop_collection(coll)
+
+
+async def remove_all_users(db: AsyncDatabase):
+    await db.command("dropAllUsersFromDatabase", 1, writeConcern={"w": async_client_context.w})
+
+
+# Constants for run_threads and lazy_client_trial.
+NTRIALS = 5
+NTHREADS = 10
+
+
+def run_threads(collection, target):
+    """Run a target function in many threads.
+
+    target is a function taking a Collection and an integer.
+    """
+    threads = []
+    for i in range(NTHREADS):
+        bound_target = partial(target, collection, i)
+        threads.append(threading.Thread(target=bound_target))
+
+    for t in threads:
+        t.start()
+
+    for t in threads:
+        t.join(60)
+        assert not t.is_alive()
+
+
+@contextlib.contextmanager
+def frequent_thread_switches():
+    """Make concurrency bugs more likely to manifest."""
+    interval = sys.getswitchinterval()
+    sys.setswitchinterval(1e-6)
+
+    try:
+        yield
+    finally:
+        sys.setswitchinterval(interval)
+
+
+def lazy_client_trial(reset, target, test, get_client):
```

Comment:
Does this even make sense to test in async? I would expect an async client to only be useable from a single thread since an I/O loop only runs on a single thread.

Reply:
It isn't being tested at all in async, not sure why I moved it here. Putting back into `test/utils.py`.